src/http/headers.rs

```diff
@@ -0,0 +1,137 @@
+use crate::ffi::{
+    ngx_http_headers_in_t, ngx_http_headers_out_t, ngx_list_part_t, ngx_list_t, ngx_pool_t, ngx_table_elt_t,
+    ngx_uint_t, u_char,
+};
+use crate::ngx_string;
+use std::ffi::CString;
+use std::fmt;
+use std::ptr;
+
+pub struct Header(*mut ngx_table_elt_t, *mut ngx_pool_t);
+
+impl Header {
+    pub fn set_value(&mut self, value: &str) {
+        // we can use pool to allocate and then copy
+        unsafe { self.0.as_mut() }.map(|table| {
+            table.value.len = value.len() as _;
+            table.value.data = crate::ffi::str_to_uchar(self.1, value);
+        });
+
+        // Alternative way is using CString and transfer ownership.
+        // unsafe { self.0.as_mut() }.map(|table| {
+        //     let c_value = CString::new(value).unwrap();
+        //     // table.value.len = c_value.as_bytes().len();
+        //     table.value.len = c_value.len();
+        //     table.value.data = c_value.into_raw() as *mut u_char;
+        // });
+    }
+
+    pub fn set_key(&mut self, key: &str) {
+        // we can use pool to allocate and then copy
+        unsafe { self.0.as_mut() }.map(|table| {
+            table.key.len = key.len() as _;
+            table.key.data = crate::ffi::str_to_uchar(self.1, key);
+            table.lowcase_key = crate::ffi::str_to_uchar(self.1, key.to_lowercase().as_str());
+        });
+    }
+
+    pub fn set_hash(&mut self, hash: ngx_uint_t) {
+        unsafe { self.0.as_mut() }.map(|table| {
+            table.hash = hash;
+        });
+    }
+
+    pub fn value(&self) -> &str {
+        unsafe {
+            std::str::from_utf8(std::slice::from_raw_parts(
+                unsafe { *self.0 }.value.data,
+                unsafe { *self.0 }.value.len as usize,
+            ))
+            .unwrap()
+        }
+    }
+
+    pub fn key(&self) -> &str {
```

Comment:
This looks like a collections (map) API. Is there a collection trait we can implement for a consistent Rust idiomatic API?

Does this follow NGINX patterns?

I think this should either be clearly recognizable as a Rust set of functions or an NGINX set of functions, my opinion leans more towards something Rust-like, especially if there's a trait.

Reply:
there is no trait in Rust. there is `pub struct HeaderMap<T = HeaderValue> { /* private fields */ }` which is here. and this struct has some methods. see: 
https://docs.rs/http/latest/http/header/struct.HeaderMap.html#

Unfortunately, we are constrained with NGINX types and implementations.

so I'm following Nginx way. in Rust's  `Header` is an abstraction that encapsulates table_elt and pool: `pub struct Header(*mut ngx_table_elt_t, *mut ngx_pool_t);` and it represents a single Header (a pair of key and value). 
in nginx you don't remove headers, you set hash to `0` ( tl;dr - headers are lists of lists ).

the next thing is to have a HeaderMap (constructed from `headers_in.headers` or `headers_out.headers` which has a set of methods similar to hyperium's`http` crate. 

anyways, seems like what I've done here at all too. if you have some ideas please share